lib/pos_snapshot_entries.go

```diff
@@ -0,0 +1,643 @@
+	}
+	return nil
+}
+
+func (bav *UtxoView) GetSnapshotValidatorByPKID(pkid *PKID, snapshotAtEpochNumber uint64) (*ValidatorEntry, error) {
+	// Check the UtxoView first.
+	mapKey := SnapshotValidatorMapKey{SnapshotAtEpochNumber: snapshotAtEpochNumber, ValidatorPKID: *pkid}
+	if validatorEntry, exists := bav.SnapshotValidatorEntries[mapKey]; exists {
+		return validatorEntry, nil
+	}
+	// If we don't have it in the UtxoView, check the db.
+	validatorEntry, err := DBGetSnapshotValidatorByPKID(bav.Handle, bav.Snapshot, pkid, snapshotAtEpochNumber)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"GetSnapshotValidatorByPKID: problem retrieving ValidatorEntry from db: ",
+		)
+	}
+	if validatorEntry != nil {
+		// Cache the result in the UtxoView.
+		bav._setSnapshotValidatorEntry(validatorEntry, snapshotAtEpochNumber)
+	}
+	return validatorEntry, nil
+}
+
+func (bav *UtxoView) GetSnapshotTopActiveValidatorsByStake(limit uint64, snapshotAtEpochNumber uint64) ([]*ValidatorEntry, error) {
+	var utxoViewValidatorEntries []*ValidatorEntry
+	for mapKey, validatorEntry := range bav.SnapshotValidatorEntries {
+		if mapKey.SnapshotAtEpochNumber == snapshotAtEpochNumber {
+			utxoViewValidatorEntries = append(utxoViewValidatorEntries, validatorEntry)
+		}
+	}
+	// Pull top N active ValidatorEntries from the database (not present in the UtxoView).
+	// Note that we will skip validators that are present in the view because we pass
+	// utxoViewValidatorEntries to the function.
+	dbValidatorEntries, err := DBGetSnapshotTopActiveValidatorsByStake(
+		bav.Handle, bav.Snapshot, limit, snapshotAtEpochNumber, utxoViewValidatorEntries,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "GetSnapshotTopActiveValidatorsByStake: error retrieving entries from db: ")
+	}
+	// Cache top N active ValidatorEntries from the db in the UtxoView.
+	for _, validatorEntry := range dbValidatorEntries {
+		// We only pull ValidatorEntries from the db that are not present in the
+		// UtxoView. As a sanity check, we double-check that the ValidatorEntry
+		// is not already in the UtxoView here.
+		mapKey := SnapshotValidatorMapKey{
+			SnapshotAtEpochNumber: snapshotAtEpochNumber, ValidatorPKID: *validatorEntry.ValidatorPKID,
+		}
+		if _, exists := bav.SnapshotValidatorEntries[mapKey]; !exists {
+			bav._setValidatorEntryMappings(validatorEntry)
+		}
+	}
+	// Pull !isDeleted, active ValidatorEntries from the UtxoView with stake > 0.
+	var validatorEntries []*ValidatorEntry
+	for mapKey, validatorEntry := range bav.SnapshotValidatorEntries {
+		if mapKey.SnapshotAtEpochNumber == snapshotAtEpochNumber &&
+			!validatorEntry.isDeleted &&
+			validatorEntry.Status() == ValidatorStatusActive &&
+			!validatorEntry.TotalStakeAmountNanos.IsZero() {
+			validatorEntries = append(validatorEntries, validatorEntry)
+		}
+	}
+	// Sort the ValidatorEntries DESC by TotalStakeAmountNanos.
+	sort.Slice(validatorEntries, func(ii, jj int) bool {
```

Comment:
we should probably be using sort.SliceStable - we also probably need a tie breaker if two validator have the same TotalStakeAmountNanos so we don't have non-determinism issues in the future.